src/plots/plots.js

```diff
@@ -865,17 +763,15 @@ plots.supplyLayoutModuleDefaults = function(layoutIn, layoutOut, fullData) {
         }
     }
 
-    // TODO register these
-    // Legend must come after traces (e.g. it depends on 'barmode')
-    var moduleLayoutDefaults = [
-        'Fx', 'Annotations', 'Shapes', 'Legend', 'Images', 'UpdateMenus'
-    ];
+    // should FX be a component?
+    Plotly.Fx.supplyLayoutDefaults(layoutIn, layoutOut, fullData);
 
-    for(i = 0; i < moduleLayoutDefaults.length; i++) {
-        _module = moduleLayoutDefaults[i];
+    var components = Object.keys(Registry.componentsRegistry);
+    for(i = 0; i < components.length; i++) {
+        _module = Registry.componentsRegistry[components[i]];
 
-        if(Plotly[_module]) {
-            Plotly[_module].supplyLayoutDefaults(layoutIn, layoutOut, fullData);
+        if(_module.supplyLayoutDefaults) {
```

Comment:
**IMPORTANT** If a registered component modules index is declared with a `supplyLayoutDefaults` method, it will get called here, automatically.

**BUT** not all registered component modules should get call here. The default step of some component modules (e.g. range sliders and range selectors) are nested inside other modules. In this case, the default handle should be exported as `handleDefaults` instead of `supplyLayoutDefaults`.

Alternatively, we could add a component module property similar to trace module `'category'`. I'm open to suggestions.